toolkit/tools/graphanalytics/graphanalytics.go

```diff
@@ -0,0 +1,287 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package main
+
+import (
+	"os"
+	"path/filepath"
+	"sort"
+
+	"gonum.org/v1/gonum/graph"
+	"gonum.org/v1/gonum/graph/traverse"
+	"gopkg.in/alecthomas/kingpin.v2"
+	"microsoft.com/pkggen/internal/exe"
+	"microsoft.com/pkggen/internal/logger"
+	"microsoft.com/pkggen/internal/pkggraph"
+	"microsoft.com/pkggen/internal/sliceutils"
+)
+
+const (
+	defaultMaxResults = "10"
+)
+
+// mapPair represents a key/value pair in a map[string][]string.
+// What the key and value represent are defined by the functions that use this.
+type mapPair struct {
+	key   string
+	value []string
+}
+
+var (
+	app            = kingpin.New("graphanalytics", "A tool to print analytics of a given dependency graph.")
+	inputGraphFile = exe.InputFlag(app, "Path to the DOT graph file to analyze.")
+	maxResults     = app.Flag("max-results", "The number of results to print per catagory. Set 0 to print unlimited.").Default(defaultMaxResults).Int()
+	logFile        = exe.LogFileFlag(app)
+	logLevel       = exe.LogLevelFlag(app)
+)
+
+func main() {
+	app.Version(exe.ToolkitVersion)
+	kingpin.MustParse(app.Parse(os.Args[1:]))
+
+	logger.InitBestEffort(*logFile, *logLevel)
+
+	err := analyzeGraph(*inputGraphFile, *maxResults)
+	if err != nil {
+		logger.Log.Fatalf("Unable to analyze dependency graph, error: %s", err)
+	}
+}
+
+// analyzeGraph analyzes and prints various attributes of a graph file.
+func analyzeGraph(inputFile string, maxResults int) (err error) {
+	pkgGraph := pkggraph.NewPkgGraph()
+	err = pkggraph.ReadDOTGraphFile(pkgGraph, inputFile)
+	if err != nil {
+		return
+	}
+
+	printDirectlyMostUnresolved(pkgGraph, maxResults)
```

Comment:
+1: I think it would help to get a more robust explanation about the "directly" and "indirectly" parts.

---
In reply to: [496090990](https://github.com/microsoft/CBL-Mariner/pull/141#discussion_r496090990) [](ancestors = 496090990)